### ISSUE TITLE
`cudaHostRegister()` may reduce time for CPU to GPU data transfer

### ISSUE THREAD

**Issue by takagi** [cupy/cupy#3450]
Currently, CuPy transfers data from CPU to GPU via pinned staging memory. I found that temporary copy to the staging area has unignorable overhead, and using `cudaHostRegister()` to page-lock the data itself may reduce the overall time to send.

In the table below, `pinned_memory` means the result to transfer via pinned staging memory, and `cudaHostRegister` to transfer with pinning the data itself. `sync` means synchronous transfer using `MemoryPointer.copy_from_host()` in reference.

|impl|size|time|
|---|---|---:|
|staging_memory|1GB|  146.299 ms|
|staging_memory|2GB|  292.464 ms|
|staging_memory|4GB|  585.627 ms|
|staging_memory|8GB| 1171.858 ms|
|cudaHostRegister|1GB|   59.668 ms|
|cudaHostRegister|2GB|  121.367 ms|
|cudaHostRegister|4GB|  251.200 ms|
|cudaHostRegister|8GB|  504.465 ms|
|sync|1GB|  149.328 ms|
|sync|2GB|  298.773 ms|
|sync|4GB|  597.199 ms|
|sync|8GB|  699.736 ms|

**Barnch to run the following script**
https://github.com/takagi/cupy/tree/pinned-memory

**Script to get the result above**
```Python
import numpy
import cupy
from cupyx import time

result = []

for impl in ['staging_memory', 'cudaHostRegister', 'sync']:
    for i in [1, 2, 4, 8]:
        a = numpy.zeros((i, 1024, 1024, 1024), dtype=numpy.uint8)
        cupy.core.core._foo_impl = impl
        x = time.repeat(cupy.array, (a,), n_repeat=5, n_warmup=1)
        result.append((impl, i, x))

print("|impl|size|time|")
print("|---|---|---:|")
for (impl, i, x) in result:
    t_ms = x.cpu_times * 1e3
    print("|{}|{}GB|{:9.03f} ms|".format(impl, i, t_ms.mean()))
```

**Comment by takagi** [comment]
Staging memory may be necessary to keep off releasing data before transfer completed, though.

**Comment by emcastillo** [comment]
> Staging memory may be necessary to keep off releasing data before transfer completed, though.

Yes, the numpy buffer might be deallocated in the middle of the data transfer. I think in the pinned you want to add an explicit reference to the a_cpu object so that the buffer is not deallocated in numpy object destruction?

**Comment by takagi** [comment]
Sounds good, I'll make a change in my branch.

**Comment by emcastillo** [comment]
Awesome! :D

**Comment by leofang** [comment]
Not sure how I missed this! Is there any plan to move forward with this, maybe provide an options for users to opt in the register approach?
